<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>exceRpt by gersteinlab</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">exceRpt</h1>
      <h2 class="project-tagline">The extra-cellular RNA processing toolkit.  Includes software to preprocess, align, quantitate, and normalise smallRNA-seq datasets</h2>
      <a href="https://github.com/gersteinlab/exceRpt" class="btn">View on GitHub</a>
      <a href="https://github.com/gersteinlab/exceRpt/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/gersteinlab/exceRpt/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span class="octicon octicon-link"></span></a>Overview</h1>

<p>The <strong>ex</strong>tra-<strong>ce</strong>llular <strong>R</strong>NA <strong>p</strong>rocessing <strong>t</strong>oolkit (<strong>exceRpt</strong>) was designed to handle the variable contamination and often poor quality data obtained from low input <strong>smallRNA-seq</strong> samples such as those obtained from <strong>extra-cellular</strong> preparations.  However the tool is perfectly capable of processing data from more standard <strong>cellular</strong> preparations and, with minor modifications to the command-line call, is also capable of processing <strong>WGS/exome</strong> and <strong>long RNA-seq</strong> data.</p>

<p>The schematic diagram below details the various stages of the exceRpt pipeline including <strong>pre-processing</strong> of the raw reads, <strong>contaminant removal</strong>, alignment to the <strong>endogenous</strong> (human/mouse) genome and annotated transcriptome, and finally alignment to a large number of <strong>exogenous</strong> sequences for detection of foreign nucleic acids.</p>

<p><img src="https://dl.dropboxusercontent.com/u/2696693/exceRpt_workflow.png" alt="exceRpt workflow diagram"></p>

<hr>

<h1>
<a id="how-to-use-excerpt" class="anchor" href="#how-to-use-excerpt" aria-hidden="true"><span class="octicon octicon-link"></span></a>How to use exceRpt</h1>

<h2>
<a id="genboreeorg" class="anchor" href="#genboreeorg" aria-hidden="true"><span class="octicon octicon-link"></span></a>Genboree.org</h2>

<p>Signing up for an account on <a href="http://genboree.org">Genboree.org</a> is by far the fastest way to start using exceRpt.  The Genboree Workbench provides a graphical interface to exceRpt that is free for academic use.  Set up a free account to use the <a href="http://genboree.org/java-bin/login.jsp">Genboree Workbench</a>, upload some data, and start using exceRpt, all via the browser (<a href="http://genboree.org/theCommons/projects/exrna-tools-may2014/wiki/Small%20RNA-seq%20Pipeline">instructions available here</a>).</p>

<h2>
<a id="docker" class="anchor" href="#docker" aria-hidden="true"><span class="octicon octicon-link"></span></a>Docker</h2>

<p>Slightly more effort to get set up than using Genboree, but using the <a href="https://www.docker.com">Docker</a> image is the best way to use exceRpt on your own machine/cluster.  The <a href="https://hub.docker.com/r/rkitchen/excerpt">exceRpt Docker image</a> is pre-baked with all required dependencies and should 'just work' with minimal installation.  Docker is native on Red Hat Enterprise Linux distributions (RHEL) and also available on Amazon, enabling exceRpt to be run in the cloud.  Instructions are available for installing Docker on <a href="http://docs.docker.com/mac/started">Mac OS X</a>, <a href="http://docs.docker.com/linux/started">Linux</a>, <a href="http://docs.docker.com/windows/started">Windows</a>, and <a href="http://docs.aws.amazon.com/AmazonECS/latest/developerguide/docker-basics.html">Amazon Web Services</a>. </p>

<p><strong>BEWARE</strong>, exceRpt requires about <strong>10GB of RAM</strong> in order to function properly, runs on machines with less than this amount of available memory will very likely fail.  If running Docker inside a virtual machine (such as via VirtualBox on Mac OS X) it is <strong>important to check that the VM has been initialised with at least 10GB of memory</strong>.  This is usually a setting that can be modified as long as the VM is not currently running.</p>

<p>As with the manual installation described below, <strong>you will also need to download the pre-compiled genome and transcriptome indices</strong> for exceRpt to use.  These are too large to be included in the Docker image, but using these links you can download the exceRpt database files:</p>

<ul>
<li>
<a href="http://homes.gersteinlab.org/people/rrk24/exceRpt/exceRptDB_hg19.tgz">human hg19</a>; </li>
<li>
<a href="http://homes.gersteinlab.org/people/rrk24/exceRpt/exceRptDB_hg38.tgz">human hg38</a>;</li>
<li>
<a href="http://homes.gersteinlab.org/people/rrk24/exceRpt/exceRptDB_mm10.tgz">mouse mm10</a>.</li>
</ul>

<p>Once you have downloaded the desired genome database and have successfully installed and started Docker on your machine/cluster/AWS-instance only a few commands are required to install and run exceRpt.  </p>

<p>To download the exceRpt Docker image, type:</p>

<pre><code>docker pull rkitchen/excerpt
</code></pre>

<p>To download the exceRpt database for your genome/species of interest (currently one of hg19, hg39, or mm10) use the following commands.  Here we'll download the latest build of the human genome and transcriptome (hg38):</p>

<pre><code>mkdir ~/DirectoryContainingMyexceRptDatabase
cd ~/DirectoryContainingMyexceRptDatabase
wget http://homes.gersteinlab.org/people/rrk24/exceRpt/exceRptDB_hg38.tgz
tar -xvf exceRptDB_hg38.tgz
</code></pre>

<p>To run a smallRNA-seq sample through the exceRpt pipeline we can grab a publicly-available dataset from the SRA:</p>

<pre><code>mkdir ~/DirectoryContainingMyInputSample
cd ~/DirectoryContainingMyInputSample
wget ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/sralite/ByExp/litesra/SRX/SRX010/SRX010851/SRR026761/SRR026761.sra
</code></pre>

<p>Almost done, just need to make sure there is a directory in which to put the exceRpt pipeline output:</p>

<pre><code>mkdir ~/DirectoryInWhichToPutMyResults
</code></pre>

<p>Finally, we can input this sample to the exceRpt pipeline, aligning to the hg38 genome+transcriptome we just downloaded:</p>

<pre><code>docker run -v ~/DirectoryContainingMyInputSample:/exceRptInput \
           -v ~/DirectoryInWhichToPutMyResults:/exceRptOutput \
           -v ~/DirectoryContainingMyexceRptDatabase/hg38:/exceRpt_DB/hg38 \
           -t rkitchen/excerpt \
           INPUT_FILE_PATH=/exceRptInput/SRR026761.sra
</code></pre>

<p>Most of the Docker command is loading directories on your machine (the <code>-v</code> parameters) so that exceRpt can read from or write to them.  The directory to the left of each <code>:</code> can obviously be whatever you want, but it is important to make sure the right side of each <code>:</code> is written as above or exceRpt will not be able to find/write the data it needs.</p>

<p>The last line of the <code>docker run</code> command (<code>INPUT_FILE_PATH=</code>) is a direct call to exceRpt itself.  As such, any customisable options should be specified this way.  For example we could re-order the priority of the transcriptome libraries (for example if we expected a sample to have more long-RNA fragments than miRNAs) using the following call to docker:</p>

<pre><code>docker run -v ~/DirectoryContainingMyInputSample:/exceRptInput \
           -v ~/DirectoryInWhichToPutMyResults:/exceRptOutput \
           -v ~/DirectoryContainingMyexceRptDatabase/hg38:/exceRpt_DB/hg38 \
           -t rkitchen/excerpt \
           INPUT_FILE_PATH=/exceRptInput/SRR026761.sra \
           ENDOGENOUS_LIB_PRIORITY=gencode,miRNA,tRNA,piRNA,circRNA
</code></pre>

<p>We can also use this approach to change the genome to which we are aligning (for example, change to mouse mm10) </p>

<pre><code>docker run -v ~/DirectoryContainingMyInputSample:/exceRptInput \
           -v ~/DirectoryInWhichToPutMyResults:/exceRptOutput \
           -v ~/DirectoryContainingMyexceRptDatabase/mm10:/exceRpt_DB/mm10 \
           -t rkitchen/excerpt \
           INPUT_FILE_PATH=/exceRptInput/SRR026761.sra \
           MAIN_ORGANISM_GENOME_ID=mm10
</code></pre>

<p>Finally, if you're lucky enough to have an abundance of available disk space, it is easier in the long run to pass Docker the directory containing all available exceRpt genome databases so that to change organism/genome only the <code>MAIN_ORGANISM_GENOME_ID</code> variable needs to be modified:</p>

<pre><code>docker run -v ~/DirectoryContainingMyInputSample:/exceRptInput \
           -v ~/DirectoryInWhichToPutMyResults:/exceRptOutput \
           -v ~/DirectoryContainingMyexceRptDatabase:/exceRpt_DB \
           -t rkitchen/excerpt \
           INPUT_FILE_PATH=/exceRptInput/SRR026761.sra \
           MAIN_ORGANISM_GENOME_ID=mm10
</code></pre>

<h2>
<a id="analysis-customisation-options" class="anchor" href="#analysis-customisation-options" aria-hidden="true"><span class="octicon octicon-link"></span></a>Analysis customisation options</h2>

<p>There are a number of options available for customising the analysis that are specified using the command-line.  These are a list of the most commonly-modified options and their default values.</p>

<pre><code>OUTPUT_DIR                  [default 'NULL']
INPUT_FILE_PATH             [default 'NULL']
CALIBRATOR_LIBRARY          [default 'NULL'] [path to bowtie2 index containing spiked-in sequences]
RANDOM_BARCODE_LENGTH       [default '0'] [options: any integer. '0' disables random barcode removal]
RANDOM_BARCODE_LOCATION     [default '-5p -3p'] [options: '-5p', '-3p', or '-5p -3p']
ENDOGENOUS_LIB_PRIORITY     [default 'miRNA,tRNA,piRNA,gencode,circRNA']
MAIN_ORGANISM_GENOME_ID     [default 'hg38'] [options: 'hg19', 'hg38', or 'mm10']
ADAPTER_SEQ                 [default 'guessKnown'] [options: 'none', 'guessKnown', or the actual nucleotide sequence of the 3' adapter]
MAP_EXOGENOUS               [default 'off'] [options: 'off', 'miRNA', or 'on']
N_THREADS                   [default '4']
</code></pre>

<h2>
<a id="manual-installation" class="anchor" href="#manual-installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Manual installation</h2>

<p>This is the most labour-intensive mode of using exceRpt, but it offers the most flexibility.  Generally not recommended unless you are very familiar with the *NIX environment, shell scripting, and makefile syntax.  Detailed instructions on how to install exceRpt and its various dependencies will [one day] be listed toward the bottom of this page.</p>

<p>Once all dependencies are installed, the exceRpt pipeline can be run on a human sample (here an SRA binary file) using default parameters using the following skeleton command:</p>

<pre><code>make -f smallRNA_pipeline.mk \
    INPUT_FILE_PATH=~/DirectoryContainingMyInputSample/SRR026761.sra \
    OUTPUT_DIR=~/DirectoryInWhichToPutMyResults
</code></pre>

<p>It is important to properly configure the makefile for this to work.  It is probably a good idea to get in touch with the maintainers using the email address listed at the bottom of the page before embarking on this yourself!</p>

<hr>

<h1>
<a id="processing-excerpt-output-from-multiple-samples" class="anchor" href="#processing-excerpt-output-from-multiple-samples" aria-hidden="true"><span class="octicon octicon-link"></span></a>Processing exceRpt output from multiple samples</h1>

<p>Also provided is a script to combine output from multiple samples run through the exceRpt pipeline.  The script (<code>mergePipelineRuns.R</code>) will take as input a directory containing 1 or more subdirectories or zipfiles containing output from the makefile above. In this way, results from 1 or more smallRNA-seq samples can be combined, several QC plots are generated, and the read-counts are normalised ready for downstream analysis by clustering and/or differential expression.</p>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>This script is comparatively much simpler to install.  Once the R software (<a href="http://cran.r-project.org/">http://cran.r-project.org/</a>) is set up on your system the script should automatically identify and install all required dependencies.  Again, this script is available on the Genboree Workbench (<a href="http://www.genboree.org">www.genboree.org</a>) and is also free for academic use.</p>

<h2>
<a id="using-the-script" class="anchor" href="#using-the-script" aria-hidden="true"><span class="octicon octicon-link"></span></a>Using the script</h2>

<h3>
<a id="on-the-command-line" class="anchor" href="#on-the-command-line" aria-hidden="true"><span class="octicon octicon-link"></span></a>On the command line</h3>

<p>The Rscript application provides a neat means of running the script:</p>

<pre><code>Rscript mergePipelineRuns.R ~/DirectoryContainingMyexceRptRuns
</code></pre>

<p>Here, the directory in which the script will search for valid exceRpt output (<code>~/DirectoryContainingMyexceRptRuns</code>) should be the same as the (<code>~/DirectoryInWhichToPutMyResults</code>) specific when running each sample through exceRpt.</p>

<h3>
<a id="interactively-in-r" class="anchor" href="#interactively-in-r" aria-hidden="true"><span class="octicon octicon-link"></span></a>Interactively in R</h3>

<p>Alternatively in an interactive R session, the merge can be performed using the following two commands:</p>

<pre><code>&gt; source("mergePipelineRuns_functions.R")
&gt; processSamplesInDir(data.dir, output.dir)
</code></pre>

<p>Apart from some status messages, warnings, or possibly errors, no R objects are output from this function.  Instead several files are created that are described immediately below...</p>

<h2>
<a id="script-output" class="anchor" href="#script-output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Script output</h2>

<p>Several files are output by the script in the location of the input exceRpt results (or somewhere else if explicitly specified).  All output files are prefixed with 'exceRpt_' and contain a variety of information regarding all samples input:</p>

<table>
<thead>
<tr>
<th>File Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>QC data:</strong></td>
<td></td>
</tr>
<tr>
<td><code>exceRpt_DiagnosticPlots.pdf</code></td>
<td>All <strong>diagnostic plots</strong> automatically generated by the merge script</td>
</tr>
<tr>
<td><code>exceRpt_readMappingSummary.txt</code></td>
<td>
<strong>Read-alignment summary</strong> including total counts for each library</td>
</tr>
<tr>
<td><code>exceRpt_ReadLengths.txt</code></td>
<td>
<strong>Read-lengths</strong> (after 3' adapters/barcodes are removed)</td>
</tr>
<tr>
<td><strong>Raw transcriptome quantifications:</strong></td>
<td></td>
</tr>
<tr>
<td><code>exceRpt_miRNA_ReadCounts.txt</code></td>
<td>
<strong>miRNA</strong> read-counts quantifications</td>
</tr>
<tr>
<td><code>exceRpt_tRNA_ReadCounts.txt</code></td>
<td>
<strong>tRNA</strong> read-counts quantifications</td>
</tr>
<tr>
<td><code>exceRpt_piRNA_ReadCounts.txt</code></td>
<td>
<strong>piRNA</strong> read-counts quantifications</td>
</tr>
<tr>
<td><code>exceRpt_gencode_ReadCounts.txt</code></td>
<td>
<strong>gencode</strong> read-counts quantifications</td>
</tr>
<tr>
<td><code>exceRpt_circularRNA_ReadCounts.txt</code></td>
<td>
<strong>circularRNA</strong> read-count quantifications</td>
</tr>
<tr>
<td><strong>Normalised transcriptome quantifications:</strong></td>
<td></td>
</tr>
<tr>
<td><code>exceRpt_miRNA_ReadsPerMillion.txt</code></td>
<td>
<strong>miRNA</strong> RPM quantifications</td>
</tr>
<tr>
<td><code>exceRpt_tRNA_ReadsPerMillion.txt</code></td>
<td>
<strong>tRNA</strong> RPM quantifications</td>
</tr>
<tr>
<td><code>exceRpt_piRNA_ReadsPerMillion.txt</code></td>
<td>
<strong>piRNA</strong> RPM quantifications</td>
</tr>
<tr>
<td><code>exceRpt_gencode_ReadsPerMillion.txt</code></td>
<td>
<strong>gencode</strong> RPM quantifications</td>
</tr>
<tr>
<td><code>exceRpt_circularRNA_ReadsPerMillion.txt</code></td>
<td>
<strong>circularRNA</strong> RPM quantifications</td>
</tr>
<tr>
<td><strong>R objects:</strong></td>
<td></td>
</tr>
<tr>
<td><code>exceRpt_smallRNAQuants_ReadCounts.RData</code></td>
<td>
<strong>All raw data</strong> (binary R object)</td>
</tr>
<tr>
<td><code>exceRpt_smallRNAQuants_ReadsPerMillion.RData</code></td>
<td>
<strong>All normalised data</strong> (binary R object)</td>
</tr>
</tbody>
</table>

<hr>

<h1>
<a id="manual-installation-1" class="anchor" href="#manual-installation-1" aria-hidden="true"><span class="octicon octicon-link"></span></a>Manual Installation</h1>

<h2>
<a id="dependencies" class="anchor" href="#dependencies" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dependencies</h2>

<h3>
<a id="core-dependencies" class="anchor" href="#core-dependencies" aria-hidden="true"><span class="octicon octicon-link"></span></a>Core dependencies</h3>

<ul>
<li><a href="https://java.com/en/download">Java v1.6+</a></li>
<li><a href="http://hannonlab.cshl.edu/fastx_toolkit">fastx v0.0.14+</a></li>
<li><a href="http://bowtie-bio.sourceforge.net/index.shtml">bowtie v1.1.1+</a></li>
<li><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2 v2.2.4+</a></li>
<li><a href="http://sourceforge.net/projects/samtools/files/samtools">samtools v1.1+</a></li>
<li><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc">fastQC v0.11.2+</a></li>
</ul>

<h3>
<a id="optional-dependencies" class="anchor" href="#optional-dependencies" aria-hidden="true"><span class="octicon octicon-link"></span></a>Optional dependencies</h3>

<ul>
<li>
<a href="https://github.com/ncbi/sra-tools">sratoolkit v2.5.1+</a> - required for processing .sra files from NCBI archival data</li>
<li>
<a href="https://github.com/alexdobin/STAR/releases">STAR v2.4.2a+</a> - required for most exogenous alignments</li>
</ul>

<h2>
<a id="installation-steps-incomplete" class="anchor" href="#installation-steps-incomplete" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation steps (incomplete)</h2>

<ol>
<li>Install dependencies</li>
<li>Download exceRpt database(s)</li>
<li>
<p>Configure the makefile, especially the following variable:</p>

<p>EXE_DIR := [location of the pipeline, database, and dependencies]</p>
</li>
</ol>

<hr>

<h1>
<a id="support" class="anchor" href="#support" aria-hidden="true"><span class="octicon octicon-link"></span></a>Support</h1>

<p>Rob Kitchen; rob [dot] kitchen [at] yale [dot] edu</p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/gersteinlab/exceRpt">exceRpt</a> is maintained by <a href="https://github.com/gersteinlab">gersteinlab</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-69502165-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
